python/ray/data/dataset.py

```diff
@@ -3553,6 +3554,68 @@ def _divide(self, block_idx: int) -> ("Dataset[T]", "Dataset[T]"):
         )
         return l_ds, r_ds
 
+    def native_batch_format(self) -> Type:
+        """Return this dataset's native batch format.
+
+        The native batch format describes what batches of data look like. To learn more
+        about batch formats, read
+        :ref:`writing user-defined functions <transform_datasets_writing_udfs>`.
+
+        Example:
+
+            If your dataset represents a list of Python objects, then the native batch
+            format is ``list``.
+
+            >>> ds = ray.data.range(100)
+            >>> ds
+            Dataset(num_blocks=20, num_rows=100, schema=<class 'int'>)
+            >>> ds.native_batch_format()
+            <class 'list'>
+            >>> next(ds.iter_batches(batch_size=4))
+            [0, 1, 2, 3]
+
+            If your dataset contains a single column of ``TensorDtype`` or
+            ``ArrowTensorType``, then the native batch format is ``ndarray``.
+
+            >>> ds = ray.data.range_tensor(100)
+            >>> ds
+            Dataset(num_blocks=20, num_rows=100, schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)})
+            >>> ds.native_batch_format()
+            <class 'numpy.ndarray'>
+            >>> next(ds.iter_batches(batch_size=4))
+            array([[0],
+                   [1],
+                   [2],
+                   [3]])
+
+            If your dataset represents structured data and the native batch format isn't
+            ``ndarray``, then the native batch format is ``DataFrame``.
+
+            >>> import pandas as pd
+            >>> df = pd.DataFrame({"foo": ["a", "b"], "bar": [0, 1]})
+            >>> ds = ray.data.from_pandas(df)
+            >>> ds
+            Dataset(num_blocks=1, num_rows=2, schema={foo: object, bar: int64})
+            >>> ds.native_batch_format()
+            <class 'pandas.core.frame.DataFrame'>
+            >>> next(ds.iter_batches(batch_size=4))
+              foo  bar
+            0   a    0
+            1   b    1
+
+        .. seealso::
+
+            :meth:`~Dataset.map_batches`
+                Call this function to transform batches of data.
+
+            :meth:`~Dataset.iter_batches`
+                Call this function to iterate over batches of data.
+
+        """  # noqa: E501
+        block = ray.get(next(self._plan.execute().iter_blocks()))
+        native_batch = BlockAccessor.for_block(block).to_native()
+        return type(native_batch)
```

Comment:
This is going to be expensive, since it involves triggering execution and fetching data. Why don't we do this based upon the dataset schema, similar to `_dataset_format()`? The mapping is:
- Pandas OR Arrow schema with more than one column OR no tensor columns: `pd.DataFrame`
- Pandas OR Arrow schema with one tensor column: `np.ndarray`
- Otherwise (scalar type indicating simple block): `list`

Reply:
We could do that. 

One drawback with this approach is that it couples `native_batch_format` with `BlockAccessor.to_native`. If we change the logic for `to_native`, we need to update `Dataset.native_batch_format`; otherwise, the method will return the wrong type. With the current approach, `Dataset.native_batch_format` always returns the correct type, even if we change the `to_native` logic.

Given that `Dataset.native_batch_format` is used for inspection, I think it might be better to go with the loosely-coupled but less performant implementation.

What do you think? I don't have a strong opinion.

Reply:
I think that making `.native_batch_format()` lightweight (at most a metadata fetch only, doesn't require execution triggering) is a big enough positive that we can stomach the higher risk of a discrepancy arising, since we should be able to mostly cover the latter via in-code comments ("make sure this is consistent with `BlockAccessor.to_native()`"), creating a shared utility for determining the native batch format between `.native_batch_format()` and `BlockAccessor.to_native()`, and adding a consistency test in CI.

Reply:
Sounds good. Will update the implementation

Reply:
Fixed